desc/magnetic_fields.py

```diff
@@ -59,6 +61,40 @@ def biot_savart(eval_pts, coil_pts, current):
     return B
 
 
+def biot_savart_general(re, rs, J, dV):
+    """Biot-Savart law for arbitrary sources.
+
+    Parameters
+    ----------
+    re : ndarray, shape(n_eval_pts, 3)
+        evaluation points to evaluate B at, in cartesian.
+    rs : ndarray, shape(n_src_pts, 3)
+        source points for current density J, in cartesian.
+    J : ndarray, shape(n_src_pts, 3)
+        current density vector at source points, in cartesian.
+    dV : ndarray, shape(n_src_pts)
+        volume element at source points
+
+    Returns
+    -------
+    B : ndarray, shape(n,3)
+        magnetic field in cartesian components at specified points
+    """
+    re, rs, J, dV = map(jnp.asarray, (re, rs, J, dV))
+    assert J.shape == rs.shape
+    JdV = J * dV[:, None]
+    B = jnp.zeros_like(re)
+
+    def body(i, B):
+        r = re - rs[i, :]
+        num = jnp.cross(JdV[i, :], r, axis=-1)
+        den = jnp.linalg.norm(r, axis=-1) ** 3
+        B = B + jnp.where(den[:, None] == 0, 0, num / den[:, None])
+        return B
```

Comment:
```python
re, rs, J, dV = map(jnp.asarray, (re, rs, J, dV))
assert J.shape == rs.shape

JdV = (dV * J.T).T
dr = re - rs[:, jnp.newaxis]
num = jnp.cross(JdV, dr, axis=-1)
den = jnp.linalg.norm(dr, axis=-1) ** 3
B = 1e-7 * jnp.nansum((num.T / den).T, axis=0)
return B
```
?

I personally prefer to use transposing to to make broadcasting work and reserve creating a new axis for when a larger matrix needs to be materialized (jnp.where duplicates `den` 3 times so you need `den[:, np.newaxis]` there, but for `JdV` you can just transpose). That way it is clear when something is being vectorized (as in my creation of `dr`) and when we're just getting broadcasting to work (like creating `JdV`).

Reply:
The issue with vectorizing this is that the `dr` array you create can be pretty huge (N^2 for N src/eval pts). Doing it in a loop limits the total memory to O(N), at the expense of being slightly less efficient on GPU at lower resolution.

Reply:
One possibly in between option would be to do the loop over whichever is smaller of the source/eval pts. right now we loop over the source points, but if we have many sources and few eval pts its a lot more efficient to loop over the eval pts.

Reply:
my vote is to loop over the vectorized algorithm in chunks. if nothing else it's cooler
```python
re, rs, J, dV = map(jnp.asarray, (re, rs, J, dV))
assert J.shape == rs.shape
JdV = (dV * J.T).T
JdV = JdV[:, jnp.newaxis]
chunk_size = 30

def body(i, B):
    start = i * chunk_size
    end = min(start + chunk_size, rs.shape[0])
    # loop over vectorized algorithm in chunks to save memory
    dr = re - rs[start:end, jnp.newaxis]
    num = jnp.cross(JdV, dr, axis=-1)
    den = jnp.linalg.norm(dr, axis=-1) ** 3
    B += jnp.nansum((num.T / den).T, axis=0)
    return B

B = 1e-7 * fori_loop(0, rs.shape[0] // chunk_size, body, jnp.zeros_like(re))
return B
```

Reply:
does this work with jit/AD if the chunk size isnt an even multiple of the total number of points?

Reply:
i couldn't get it to work. if it does its not worth added code complexity